File: api/services/SyncService.js
'use strict';

const Service = require('trails/service');
const Err = require('err');
const _ = require('lodash');
const simpleGit = require('simple-git');
const fs = require('extfs');

/**
 * @module SyncService
 * @description Handles one way sync
 */
module.exports = class SyncService extends Service {

  sync(system) {
    switch(system) {
    case 'git':
      return this._gitSync();
      break;
    default:
      return Promise.reject(new Err('Invalid sync system', 400));
      break;
    }
  }

  _gitSync() {
    const blogdown = this.app.config.blogdown;
    const git = blogdown.sync.git;
    return new Promise((resolve, reject) => {
      fs.isEmpty(blogdown.contentPath, function(empty) {
        resolve(!empty);
      });
    }).then((exists) => {
      return new Promise((resolve, reject) => {
        let exec = require('child_process').exec;
        if (exists) { // pull
          exec(`git pull origin ${git.branch}`, (err, stdout) => {
            if (err) reject(err);
            resolve('Successfully pulled');
          });
        } else { // clone
          exec(`git clone ${git.repo} ${blogdown.contentPath}`, (err, stdout) => {
            if (err) reject(err);
            exec(`git checkout ${git.branch}`, (err, stdout) => {
              if (err) reject(err);
              resolve('Successfully cloned');
            });
          });
        }
      });
    }).then(result => {
      return result;
    });
  }
};

File: api/services/sync_service.py
import git
import os
import shutil
from git import Git, Repo
from api.config import config

def sync(system):
    if system == 'git':
        message = __sync_git()
        return {
            'message': message
        }
    return {
        'message': 'Invalid system'
    }

def __sync_git():
    message = ''
    repo = {}
    os.popen('''mkdir -p /tmp/backdown/content/
    mv ''' + config.contentPath + '''* /tmp/backdown/content/
    mv ''' + config.contentPath + '''.* /tmp/backdown/content/
    ''')
    try:
        repo = Repo.clone_from(config.sync['git']['repo'], '/tmp/backdown/content/')
        message = config.sync['git']['repo'] + ' cloned'
    except git.GitCommandError as e:
        if 'already exists and is not an empty directory' in e.stderr:
            repo = Repo('/tmp/backdown/content/')
            repo.remotes.origin.pull()
            message = config.sync['git']['repo'] + ' pulled'
        else:
            raise e
    os.popen('''mv /tmp/backdown/content/* ''' + config.contentPath + '''
    mv /tmp/backdown/content/.* ''' + config.contentPath + '''
    rm -rf /tmp/backdown/content/
    ''')
    return message
